up_client_socket/java/src/main/java/org/eclipse/uprotocol/SocketUTransport.java

```diff
@@ -1,274 +1,241 @@
+      * Initializes the socket and starts listening for incoming messages asynchronously.
+      *
+      * @param newSource The source URI to be used by this transport.
+      * @throws IOException If an I/O error occurs when creating the socket.
+      */
+     public SocketUTransport(UUri newSource) throws IOException {
+         source = newSource;
+         uriToListener = new HashMap<>();
+         socket = new Socket(DISPATCHER_IP, DISPATCHER_PORT);
+         ExecutorService executor = Executors.newFixedThreadPool(5);
+         executor.submit(this::listen);
+         executor.shutdown();
+     }
+ 
+ 
+     /**
+      * Listens for incoming UMessages from the Dispatcher.
+      * Processes the incoming data if the listener is registered with a UMessage source and sink UURI filter.
+      */
+     private void listen() {
+         try {
+             while (true) {
+                 byte[] buffer = new byte[BYTES_MSG_LENGTH];
+                 InputStream inputStream = socket.getInputStream();
+                 int readSize = inputStream.read(buffer);
+ 
+                 if (readSize < 0) {
+                     if (!socket.isClosed()) {
+                         socket.close();
+                     }
+                     return;
+                 }
+                 UMessage umsg = UMessage.parseFrom(Arrays.copyOfRange(buffer, 0, readSize));
+                 String logMessage = " Received uMessage";
+                 logger.info(logMessage);
+                 notifyListeners(umsg);
+             }
+         } catch (IOException e) {
+             try {
+                 if (!socket.isClosed()) {
+                     socket.close();
+                 }
+             } catch (IOException ioException) {
+                 logger.log(Level.SEVERE, String.format("Error while closing socket: %s", ioException.getMessage()),
+                         ioException);
+             }
+             logger.log(Level.SEVERE, String.format("Error while listening for messages: %s", e.getMessage()), e);
+         }
+     }
+ 
+     /**
+      * Notifies listeners registered for the source and sink URI filters about the incoming message.
+      * The message is matched against the registered URI filters, and the appropriate listeners are
+      * invoked asynchronously.
+      *
+      * @param umsg The message to be processed and dispatched to listeners.
+      */
+     public void notifyListeners(UMessage umsg) {
+         synchronized (lock) {
+             boolean isMatch = false;
+             for (Map.Entry<AbstractMap.SimpleEntry<UUri, UUri>, UListener> entry : uriToListener.entrySet()) {
+                 AbstractMap.SimpleEntry<UUri, UUri> key = entry.getKey();
+                 UListener listener = entry.getValue();
+                 UriFilter uriFilter = new UriFilter(key.getKey(), key.getValue());
+                 boolean match = uriFilter.matches(umsg.getAttributes());
```

Comment:
What happens in case only source filter is only set for register listener ? Should this be comparing source and sink individually and as pair depending on registerlistener ?

Reply:
default CompletionStage<UStatus> registerListener(UUri sourceFilter, UListener listener) {
        return registerListener(sourceFilter, UriFactory.ANY, listener);
    }
    
    When this is used, the sink filter will be set to wildcard, so it should be handled.